062-quotas-management.md

```diff
@@ -0,0 +1,167 @@
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    quotas:
+      type: kafka
+      producerByteRate: 1048576
+      consumerByteRate: 2097152
+      requestPercentage: 55
+      controllerMutationRate: 50
+    # ... 
+  # ...
+```
+
+Based on this configuration, the operator will do the following:
+* Configure the default quota plugin (make sure the `client.quota.callback.class` option is not set)
+* Set the default user quotas
+* When the `type: kafka` quotas API is set, but one or more of the properties with the individual quotas are missing, the operator will make sure that the given default quota is not set (set to `null`).
+* Other configuration options for the built-in Kafka plugin can be configured by the user in the `.spec.kafka.config` section in the `Kafka` CR.
+
+The second supported plugin - the Strimzi Quota plugin - will be configured as `type: strimzi`.
+The API will allow you to configure the storage limits at which the plugin will start limiting the producers in order to prevent the disk from getting full.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    quotas:
+      type: strimzi
+      produceQuota: 1000000
+      fetchQuota: 1000000
+      minAvailableBytesPerVolume: 5368709120
+      excludedPrincipals:
+        - my-special-user
+    # ... 
+  # ...
+```
+
+The `type: strimzi` quotas API will support the following fields:
+* `produceQuota` for setting the default produce quota
+* `fetchQuota` for setting the default fetch quota
+* `minAvailableBytesPerVolume` for setting the minimal required free space per volume as number of bytes (this field can be configured only when `minAvailableRatioPerVolume`is not set)
+* `minAvailableRatioPerVolume` for setting the minimal required free space per volume as percentage of the overall volume size (this field can be configured only when `minAvailableBytesPerVolume`is not set)
+* `excludedPrincipals` for setting a list of users to be excluded from the applied quotas
+
+Other options might be configured directly by the user in `.spec.kafka.config` section of the `Kafka` CR.
+
+Based on this configuration, the operator will:
+* Set the `client.quota.callback.class` field in Kafka configuration to `io.strimzi.kafka.quotas.StaticQuotaCallback`.
+* Set the configuration on how to connect to the Kafka broker as a client using the `client.quota.callback.static.kafka.admin.` options.
+* Configure the default produce and fetch quota (`client.quota.callback.static.produce` and `client.quota.callback.static.fetch`).
+* Configure the minimal available disk space per-volume using the `client.quota.callback.static.storage.per.volume.limit.min.available.bytes` or `client.quota.callback.static.storage.per.volume.limit.min.available.ratio` options.
```

Comment:
I guess they will not be set and the defaults from the Quotas plugin will be used. But I think it is a valid use-case that you want to just set the throughput quota and not care about the storage.

Reply:
> I guess they will not be set and the defaults from the Quotas plugin will be used.

Looks like there is no default, which I guess it means no check: https://github.com/strimzi/proposals/blob/20a4043c5a2f86e35995eb12b1ac50aac8929154/047-cluster-wide-volume-usage-quota-management.md#limit-configuration

I think it would be better to be opinionated on that an set a default value.

Reply:
I don't think so. As I said, it might be a valid situation that you just want to disable it.

Reply:
Also, if we would decide to have a default, it should be done in the Qoutas plugin and not in the operator

Reply:
> if we would decide to have a default, it should be done in the Qoutas plugin and not in the operator

I agree on this. @SamBarker wdyt?

Reply:
I sort of agree. If we are to have an opinionated default it should be done in the plug-in.

There was a lot of hesitation during the proposal review about defaulting to blocking access so we default `client.quota.callback.static.throttle.factor.fallback` to `1.0` so I'm not sure we have a different calculus here (personally I think we should default to blocking but thats another conversation). 

The bigger question is how to determine a sensible default? Setting `min.available.bytes=1GB` allows almost maximum usage but potentially makes for un-recoverable brokers. If they crash for other reasons there still isn't room on the volume to do log recovery. Using `min.available.ratio=0.05` is more forgiving but leads to questions about wasted disk space and potentially the same flaw (I don't have a picture of how much space is required for log recovery).

Reply:
Can't we figure out a way to compute this default based on the number of IO threads and the segment size (Tom's suggestion)? We may need to dig into the code to figure this out.

I've seen many users blindingly trusting Kafka and Strimzi default values in production. For this reason, I also think it's important to have a default value, even if it's not perfect.

@showuon @mimaison @tombentley what's your opinion on this?

Reply:
* We haven't been able to finish the quotas stuff and bring it to working state for several years now. Do you really want to now start creating some complex schemas how to define storage limits based on number of threads?
* I think you are still missing the main point. The quotas will not be enabled by default and they cannot be enabled by default because of Kafka's native quotas. So the default affects only people who intentionally enable the quotas but without defining the storage limit. We can discuss how big the group of people who would enable a plugin for quotas based on storage but not set the limit by mistake is. But it is certainly not equal to `users blindingly trusting Kafka and Strimzi default values in production`.

Reply:
> the default affects only people who intentionally enable the quotas but without defining the storage limit.

I'm aware, see my first comment on this thread.

> We can discuss how big the group of people who would enable a plugin for quotas based on storage but not set the limit by mistake is.

The real question is: Can you trust them setting a sensible value that would avoid the potential log recovery issue that Sam was mentioning?

Not saying we should do it now, but maybe spending some time thinking about it would be good.